include/libtorrent/ssl/context.hpp

```diff
@@ -0,0 +1,279 @@
+	{
+		set_password_callback([pass](std::size_t, password_purpose p)
+			{ return p == for_reading ? pass : ""; }
+			, ec);
+		return ec;
+	}
+
+	error_code set_verify_dn(std::string name, error_code& ec)
+	{
+		set_verify_callback(
+				std::bind(&context::verify_dn
+					, std::move(name)
+					, std::placeholders::_1
+					, std::placeholders::_2)
+				, ec);
+		return ec;
+	}
+
+	error_code set_trust_certificate(const const_buffer& certificate, file_format format, error_code &ec)
+	{
+#if defined TORRENT_USE_OPENSSL
+		if(format != pem)
+			return ec = {boost::asio::error::operation_not_supported, error::get_ssl_category()};
+
+		// create a new X.509 certificate store
+		X509_STORE* cert_store = X509_STORE_new();
+		if (!cert_store)
+			return ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+
+		// wrap the PEM certificate in a BIO, for openssl to read
+		BIO* bp = BIO_new_mem_buf(
+			const_cast<void*>(static_cast<void const*>(certificate.data()))
+			, int(certificate.size()));
+
+		// parse the certificate into OpenSSL's internal representation
+		X509* cert = PEM_read_bio_X509_AUX(bp, nullptr, nullptr, nullptr);
+		BIO_free(bp);
+
+		if (!cert)
+		{
+			X509_STORE_free(cert_store);
+			return ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		}
+
+		// add cert to cert_store
+		X509_STORE_add_cert(cert_store, cert);
+		X509_free(cert);
+
+		// and lastly, replace the default cert store with ours
+		SSL_CTX_set_cert_store(native_handle(), cert_store);
+		return ec;
+#elif defined TORRENT_USE_GNUTLS
+		std::string cert(static_cast<char const*>(certificate.data()), certificate.size());
+
+        gnutls_datum_t ca;
+        ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(cert.c_str())); // must be null terminated
+        ca.size = cert.size();
+
+		// Warning: returns the number of certificates processed or a negative error code on error
+		int ret = gnutls_certificate_set_x509_trust_mem(native_handle()
+				, &ca
+				, format == pem ? GNUTLS_X509_FMT_PEM : GNUTLS_X509_FMT_DER);
+		return ec = ret >= 0 ? error_code() : error_code(ret, error::get_ssl_category());
+#endif
+	}
+
+#if defined TORRENT_USE_OPENSSL
+	error_code set_servername_callback(std::function<bool(stream_base& s, std::string name)> cb, error_code& ec)
+	{
+#if OPENSSL_VERSION_NUMBER >= 0x90812f
+#if defined __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+		SSL_CTX_set_tlsext_servername_callback(native_handle(), (cb ? servername_func : nullptr));
+		SSL_CTX_set_tlsext_servername_arg(native_handle(), (cb ? m_impl.get() : nullptr));
+#if defined __clang__
+#pragma clang diagnostic pop
+#endif
+		m_impl->servername_callback = std::move(cb);
+		return ec;
+#else // OPENSSL_VERSION_NUMBER
+		return ec = {boost::asio::error::operation_not_supported, error::get_ssl_category()};
+#endif
+	}
+#endif
+
+private:
+	static void global_init();
```

Comment:
this `global_init()` function doesn't seem all that useful. Couldn't you just do the openSSL initialization in a global constructor, as it used to be? Now it looks like you run it every time a `context` is created which happens multiple times. It also doesn't seem to benefit from being a member function here.

Reply:
The init was done every time a torrent context was created before: https://github.com/arvidn/libtorrent/blob/48627d7dfe47fb50a99fbd429de25bebec129dfc/src/torrent.cpp#L1636
I'll move it in a global constructor just like the cleanup is in a global destructor.

Reply:
oh, I see. I think I confused it with how I used to initialize openssl a long time ago, where it was in the same global static-storage duration object that does the cleanup ( https://github.com/arvidn/libtorrent/blob/master/src/session_impl.cpp#L172 ). It's possible I moved it to be done more often for a reason. It could be risky to call it over and over again though